.github/workflows/lint-and-test.yaml

```diff
@@ -2,43 +2,186 @@ name: Lint and test PR
+          charts:
+            - 'charts/**'
 
-  lint-chart:
+
+  docker-build:
+    name: Build PowerDNS image
     runs-on: ubuntu-latest
-    needs: need-test
-    if: ${{ needs.need-test.outputs.charts == 'true' }}
+    needs: paths-filter
+    if: ${{ needs.paths-filter.outputs.build == 'true' || needs.paths-filter.outputs.workflows == 'true' }}
     steps:
-      - uses: actions/checkout@v3
-        with:
-          fetch-depth: 0
-      - uses: helm/chart-testing-action@v2.3.1
-      - run: helm plugin install https://github.com/helm-unittest/helm-unittest
-      - uses: manusa/actions-setup-minikube@v2.7.2
-        with:
-          minikube version: v1.29.0
-          kubernetes version: v1.25.0
-          github token: ${{ secrets.GITHUB_TOKEN }}
-
-      - run: ct --config .github/ct.yaml lint --debug
+    - name: Checkout code
+      uses: actions/checkout@v3
+    - name: Set up Docker Buildx
+      uses: docker/setup-buildx-action@v2
 
-  docker-build:
+    - name: Build image
+      uses: docker/build-push-action@v4
+      with:
+        context: build
+        tags: ghcr.io/txqueuelen/stateless-dns:ci
+        outputs: type=docker,dest=/tmp/stateless-dns.tar
+
+    - name: Upload artifact
+      uses: actions/upload-artifact@v3
+      with:
+        name: docker-image
+        path: /tmp/stateless-dns.tar
+
+
+  lint-charts:
+    name: Lint charts
     runs-on: ubuntu-latest
+    needs: [paths-filter, docker-build]
+    if: ${{ needs.paths-filter.outputs.charts == 'true' || needs.paths-filter.outputs.workflows == 'true' }}
     steps:
-      - uses: actions/checkout@v3
-      - run: docker build . -t pdns-stateless:ci
+    - name: Checkout code
+      uses: actions/checkout@v3
+      with:
+        fetch-depth: 0
+    - name: Install helm/chart-testing
+      uses: helm/chart-testing-action@v2.3.1
+
+    # Lint charts and run unit tests
+    - name: Lint charts
+      run: ct --config .github/ct.yaml lint --all
+
+    - name: Install helm unit test siute
+      run: helm plugin install https://github.com/helm-unittest/helm-unittest
+    - name: Run unit tests
+      run: |
+        for chart in $(ct list-changed --config .github/ct-lint.yaml); do
+          if [ -d "$chart/tests/" ]; then
+            helm unittest $chart -3
+          fi
+        done
+
+    # Download the image built before or a stable one
+    - name: Download artifact from previous job
+      if: needs.docker-build.result != 'skipped'
+      uses: actions/download-artifact@v3
+      with:
+        name: docker-image
+        path: /tmp
+    - name: Load image from previous job
+      if: needs.docker-build.result != 'skipped'
+      run: docker load --input /tmp/stateless-dns.tar
+
+    - name: Rename latest docker build as ci to make e2e
+      if: needs.docker-build.result == 'skipped'
+      run: |
+        docker pull ghcr.io/txqueuelen/stateless-dns:latest
+        docker tag ghcr.io/txqueuelen/stateless-dns:latest ghcr.io/txqueuelen/stateless-dns:ci
```

Comment:
I understand we need this complexity because docker build is hella slow here. I need to think more about this, there should be a way to make this easier with just the registry.

Reply:
As I said above:

> Github only gives us 500Mb of free package storage. I am not going to push/upload any image until I am completely sure that passes all the tests.
> 
> In fact, image and chart should be uploaded on release.

Here we are testing the charts. I want to test the chart with the latest version of the image or the one that came from build step. That is the reason this exist. Nothing related to make it fast.